fdbserver/storageserver.actor.cpp

```diff
@@ -3360,6 +3361,40 @@ bool storageServerTerminated(StorageServer& self, IKeyValueStore* persistentData
 		return false;
 }
 
+ACTOR Future<bool> memoryStoreRecover(IKeyValueStore* store, Database cx, UID id)
+{
+    if(store->getType() != KeyValueStoreType::MEMORY) {
+        return false;
+    }
+
+    state Transaction tr( cx );
+	state int noCanRemoveCount = 0;
+	loop {
+		try {
+			if (noCanRemoveCount >= 5) {
+			// retry five times, can't remove storage server, needs to wait recovery done
+			return false;
+			}
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+			state bool canRemove = wait( canRemoveStorageServer( &tr, id ) );
+			if (!canRemove) {
+				TEST(true); // it's possible that the caller had a transaction in flight that assigned keys to the server. Wait for it to reverse its mistake.
+				Void _ = wait( delayJittered(SERVER_KNOBS->REMOVE_RETRY_DELAY, TaskUpdateStorage) );
+				tr.reset();
+				TraceEvent("RemoveStorageServerRetrying").detail("Count", noCanRemoveCount++).detail("ServerID", id).detail("CanRemove", canRemove);
+			} else {
+				Void _ = wait(tr.commit());
```

Comment:
Is there a reason to call `commit` here? I don't see any write operations, so this looks like a no-op to me

Reply:
oh, from my understanding,  inside [canRemoveStorageServer](https://github.com/apple/foundationdb/blob/release-5.2/fdbserver/MoveKeys.actor.cpp#L693) method, we use the txn the read the system key.

Reply:
I'm not sure I fully understand what's going on here but it does seem that tr has a read conflict range, so committing it is not a total no-op. I don't see why it matters whether a key gets written to the read conflict range before or after the transaction commits though

Reply:
Just discussed this with @atn34. In the context of this change, the commit is a no-op. What it does is it will verify that nobody wrote to the key range in the time between the read and the commit. But since this key range only has one writer (the DD role), it probably doesn't really matter.

What you could do is the following: in the `if (!canRemove)` branch you could commit and if you get a conflict (`error_code_not_committed`) you could immediately retry. This would be a very minor optimization. But honestly I would assume that it doesn't help you much.

I would suggest to remove this commit. The only thing it seems to do is to add unnecessary load onto the resolver.

Reply:
If there isn't a write conflict range, then `commit()` is a no-op.  Committing a read-only transaction doesn't cause it to be sent to a resolver.  It's likely only being committed because committing your transactions is a good thing to do.